File: 0971-shortest-bridge/0971-shortest-bridge.cpp
class Solution {
    vector<pair<int,int>> dir = {{0,1},{0,-1},{1,0},{-1,0}}; 
    bool visited[101][101]; 
public: 
    void dfs(vector<vector<int>>& grid, int i, int j){
        if(i < 0 || j < 0 || i >= grid.size() || j >= grid[0].size() || grid[i][j] != 1) return; 

        grid[i][j] = 2; 

        dfs(grid,i+1,j); 
        dfs(grid,i-1,j);
        dfs(grid,i,j+1);
        dfs(grid,i,j-1); 
    }

public:
    int bfs(vector<vector<int>>& grid, int i, int j){
        queue<vector<int>> q; 
        q.push({i,j,0}); 
        //vector<vector<bool>> visited(grid.size(),vector<bool>(grid[0].size(),false)); 

        visited[i][j] = true; 

        while(!q.empty()){
            int size = q.size(); 
            for(int i = 0; i < size; i++){
                vector<int> first = q.front();
                q.pop(); 

                int x = first[0], y = first[1], dist = first[2]; 
                if(grid[x][y] == 2){
                    return dist -1; 
                }

                for(pair<int,int>& p : dir){
                    int nX = x + p.first; 
                    int nY = y + p.second; 

                    if(nX < 0 || nX >= grid.size() || nY < 0 || nY >= grid[0].size() || grid[nX][nY] == 1 || visited[nX][nY]) continue; 

                    visited[nX][nY] = true; 
                    q.push({nX,nY,dist+1}); 
                }
            }
        }
        return INT_MAX; 
    }
public:
    int shortestBridge(vector<vector<int>>& grid) {
        bool flag = true; 
        int answer = INT_MAX; 
        for(int i = 0; i < grid.size(); i++){
            for(int j = 0; j < grid[0].size(); j++){
                if(grid[i][j] == 1){
                    dfs(grid,i,j); 
                    flag = false; 
                    break; 
                }
            }
            if(!flag) break; 
        }
        for(int i = 0; i < grid.size(); i++){
            for(int j = 0; j < grid[i].size(); j++){
                memset(visited,false,sizeof(visited)); 
                if(grid[i][j] == 1){
                    answer = min(answer,bfs(grid,i,j)); 
                }
            }
        }


        return answer; 
    }
};

File: 0971-shortest-bridge/0971-shortest-bridge.java
class Solution {
    int[][] dir = {{0,1},{0,-1},{-1,0},{1,0}}; 
    class Node{
        int x, y, dist; 
        Node(int x, int y, int dist){
            this.x = x;
            this.y = y; 
            this.dist = dist; 
        }
    }
    public void dfs(int i, int j, int[][] grid){
        if(i < 0 || j < 0 || i >= grid.length || j >= grid[0].length || grid[i][j] != 1) return; 
        grid[i][j] = 2; 
        dfs(i+1,j,grid); 
        dfs(i-1,j,grid); 
        dfs(i,j+1,grid);
        dfs(i,j-1,grid); 
    }
    public int bfs(int i, int j, Queue<Node> q, int[][] grid){
        boolean[][] visited = new boolean[grid.length][grid[0].length]; 
        visited[i][j] = true; 
        
        while(!q.isEmpty()){
            int size = q.size(); 
            for(int k = 0; k < size; k++){
                Node first = q.poll(); 
                int x = first.x, y = first.y, dist = first.dist; 
                
                if(grid[x][y] == 2) return dist - 1; 
                
                for(int[] d : dir){
                    int nX = x + d[0]; 
                    int nY = y + d[1]; 
                    
                    if(nX < 0 || nY < 0 || nX >= grid.length || nY >= grid[0].length || visited[nX][nY] || grid[nX][nY] == 1) continue; 
                    visited[nX][nY] = true; 
                    q.add(new Node(nX,nY,dist+1)); 
                }
            }
        }
        return Integer.MAX_VALUE; 
    }
    public int shortestBridge(int[][] grid) {
        
        boolean flag = false; 
        int answer = Integer.MAX_VALUE; 
        for(int i = 0; i < grid.length; i++){
            for(int j = 0; j < grid[i].length; j++){
                if(!flag && grid[i][j] == 1){
                    dfs(i,j,grid); 
                    flag = true; 
                }
            }
            if(flag) break; 
        }
        
        for(int i = 0; i < grid.length; i++){
            for(int j = 0; j < grid[i].length; j++){
                if(grid[i][j] == 1){
                    Queue<Node> q = new LinkedList<>(); 
                    q.add(new Node(i,j,0)); 
                    answer = Math.min(answer,bfs(i,j,q,grid)); 
                }
            }
        }
        return answer; 
    }
}